libs/cardano-data/src/Data/OMap/Strict.hs

```diff
@@ -0,0 +1,360 @@
+    snoc' omap@(OMap seq kv) v =
+      let k = v ^. okeyL
+       in if Map.member k kv
+            then omap
+            else OMap (seq SSeq.|> v) (Map.insert k v kv)
+
+-- | \(O(n\log n)\). Checks membership before snoc'ing.
+-- De-duplicates the StrictSeq and collects and returns the duplicates found.
+-- Starts from the left or head, using `foldl'`
+fromStrictSeqDuplicates :: (HasOKey k v, Ord k, Ord v) => SSeq.StrictSeq v -> (Set.Set v, OMap k v)
+fromStrictSeqDuplicates = F.foldl' snoc' (Set.empty, empty)
+  where
+    snoc' :: (HasOKey k v, Ord k, Ord v) => (Set.Set v, OMap k v) -> v -> (Set.Set v, OMap k v)
+    snoc' (duplicates, omap@(OMap seq kv)) v =
+      let k = v ^. okeyL
+       in if Map.member k kv
+            then (Set.insert v duplicates, omap)
+            else (duplicates, OMap (seq SSeq.|> v) (Map.insert k v kv))
+
+-- | \(O(n\log n)\).
+fromSet :: (HasOKey k v, Ord k) => Set.Set v -> OMap k v
+fromSet = fromStrictSeq . SSeq.fromList . Set.elems
+
+-- | \(O(1)\).
+toMap :: HasOKey k v => OMap k v -> Map.Map k v
+toMap = omMap
+
+-- | \(O(1)\).
+toStrictSeq :: OMap k v -> SSeq.StrictSeq v
+toStrictSeq = omSSeq
+
+-- | \(O(n\log n)\).
+toStrictSeqOKeys :: HasOKey k v => OMap k v -> SSeq.StrictSeq k
+toStrictSeqOKeys = fmap (^. okeyL) . omSSeq
+
+-- | \(O(n\log n)\).
+toStrictSeqOfPairs :: HasOKey k v => OMap k v -> SSeq.StrictSeq (k, v)
+toStrictSeqOfPairs = fmap (\v -> (v ^. okeyL, v)) . omSSeq
+
+-- | \(O(\log n)\). Key membership check.
+member :: (HasOKey k v, Ord k) => k -> OMap k v -> Bool
+member k (OMap _seq kv) = Map.member k kv
+
+-- | \(O(\log n)\). Value membership check.
+elem :: (HasOKey k v, Ord k) => v -> OMap k v -> Bool
+elem v = member (v ^. okeyL)
+
+-- | \(O(n\log n)\)
+filter :: (HasOKey k v, Ord k) => (v -> Bool) -> OMap k v -> OMap k v
+filter f = F.foldl' combine empty
+  where
+    combine accum v =
+      if f v
+        then accum |> v
+        else accum
+
+-- | \(O(n)\). Given a `Set` of @k@s, and an `OMap` @k@ @v@ return
+-- a pair of `Map` and `OMap` where the @ks@ in the `Set` have been
+-- removed from the `OMap` and presented as a separate `Map`.
+extractKeys :: (HasOKey k v, Ord k) => Set.Set k -> OMap k v -> (Map.Map k v, OMap k v)
+extractKeys ks (OMap seq kv) =
+  let (kv', extractedKv) = MapE.extractKeys kv ks
+      seq' =
+        F.foldl'
+          (\accum v -> if Set.member (v ^. okeyL) ks then accum else accum SSeq.|> v)
+          SSeq.empty
+          seq
+   in (extractedKv, OMap seq' kv')
+
+-- | \(O(\log n)\). Like `Map.adjust`. Returns the original `OMap`
+-- unaltered when the key does not exist in the `OMap` or when the
+-- `OMap` invariant is not sustained after the operation.
+adjust :: (HasOKey k v, Ord k) => (v -> v) -> k -> OMap k v -> OMap k v
+adjust f k omap@(OMap seq kv) =
+  case Map.lookup k kv of
+    Nothing -> omap
+    Just v ->
+      let v' = f v
+       in if v' ^. okeyL == k
+            then OMap seq (Map.insert k v' kv)
+            else omap
```

Comment:
> I am pretty sure this is the reason why you are getting test failures.

Alexey! :heart: :bow: what a great catch! I was also suspecting the voting to be wrong, but I could not spot the problem. Thank you so much for spotting it!

Reply:
Given the usages for the respective operations is the following (correct me if I'm wrong @TimSheard @lehins )

Assuming `k` for keys (`GovActionId`)  and `v` for values (`GovActionState`)
1. Updates/adjustments to `v` are expected to happen quite often, given that this operation is how voting would be performed.
2. The pulsar needs to have an ordered sequence of `v`s given to its closure, only once per epoch.

Therefore, we need more efficient updates/adjustments and we _can do_ with a less efficient `O(n * log n)` ordered sequence extraction operation. 

If we keep the `OMap k v = { x :: StrictSeq v, y :: Map k v }` we get inefficient adjustments/updates since we have to keep the `v`s updated and maintain the invariant `O(n + log n)`, but efficient ordered sequence extraction for the pulsar `O(1)`.

If we revert it back to the original `OMap k v = { x :: StrictSeq k, y :: Map k v }` we get inefficient ordered sequence extraction for the pulsar `O(n)`, but efficient adjustments/updates since we have to update/adjust `v` only in the underlying `Map` in `O(log n)`, and updates and adjustments are going to be many times in an epoch.

I think you are totally correct @lehins and I would like to revert this.